Updated documentation for rootless release (#133)
AxisCommunications/docker-acap: README.md

| Release                 | AXIS OS min. version | Dockerd version | Type     | Comment                         |
| ----------------------: | -------------------: | --------------: |----------|---------------------------------|
| [2.0.N][latest-release] | 11.10                | 24.0.2          | rootless | Latest release                  |
| [1.4.1][1.4.1-release]  | 10.12                | 20.10.9         | rootful  | Legacy release AXIS OS 2022 LTS |
| 2.0.N                   | 11.11                | 24.0.2          | rootless | Legacy release AXIS OS 2024 LTS |
| 1.5.N                   | 11.11                | 24.0.2          | rootful  | Legacy release AXIS OS 2024 LTS |
<!-- markdownlint-enable MD013 -->
> [!IMPORTANT]
> From AXIS OS 12.0 running 'rootful' ACAP applications, i.e. an application setup with the `root` user,
> will no longer be supported. To install a 'rootful' ACAP application on a device running AXIS OS
> versions between 11.5 and 11.11, allow root must be enabled. See the [VAPIX documentation][vapix-allow-root]
> for details. Alternatively, On the web page of the device:
>
> 1. Go to the Apps page, toggle on `Allow root-privileged apps`.

> > 1. Go to System -> Account page, under SSH accounts toggle off `Restrict root access` to be able to

Reviewer: Should it be '2.'?

  - [Using an SD card as storage](#using-an-sd-card-as-storage)
  - [Using the Docker ACAP application](#using-the-docker-acap-application)
- [Building the Docker ACAP application](#building-the-docker-acap-application)
- [Installing a locally built Docker ACAP application](#installing-a-locally-built-docker-acap-application)
- [Contributing](#contributing)
- [License](#license)

## Overview

The Docker ACAP application provides the means to run a Docker daemon on an Axis device, thereby
making it possible to deploy and run Docker containers on it. When started the daemon
will run in rootless mode, i.e. the user owning the daemon process will not be root,
and by extension, the containers will not have root access to the host system.
See [Rootless Mode][docker-rootless-mode] on Docker.com for more information. The page also
contains known limitations when running rootless Docker.

<!-- omit in toc -->
### Known Issues

- Only uid and gid are properly mapped between device and containers, not the other groups that the
user is a member of. This means that resources on the device, even if they are volume or device mounted

> can be inaccessible inside the container. This can also affect usage of unsupported dbus methods from

Reviewer 2: ```suggestion
can be inaccessible inside the container. This can also affect usage of unsupported D-Bus methods from
```

curl -s https://api.github.com/repos/AxisCommunications/docker-acap/releases/latest | grep "browser_download_url.*Docker_Daemon_.*_<ARCH>\_signed.eap"
```

#### Migrating from rootful Docker ACAP application

When migrating from a rootful Docker ACAP application, any version before 2.0, the following is recommended:

- Copy any images that you want to persist from the device to your computer.
- Stop the Docker ACAP application.
- Uninstall the Docker ACAP application.

> - If you use the SD card as storage either format it or manually remove the `dockerd` directory (`/var/spool/storage/SD_Disk/dockerd`).

Reviewer 3: If you use the SD card as storage either format it or manually remove the `dockerd` directory (`/var/spool/storage/SD_DISK/dockerd`).
```suggestion
- If you use the SD card as storage either format it or manually remove the `dockerd` directory (`/var/spool/storage/SD_Disk/dockerd`).
```

curl -s https://api.github.com/repos/AxisCommunications/docker-acap/releases/latest | grep "browser_download_url.*Docker_Daemon_.*_<ARCH>\_signed.eap"
```

#### Migrating from rootful Docker ACAP application

When migrating from a rootful Docker ACAP application, any version before 2.0, the following is recommended:

- Copy any images that you want to persist from the device to your computer.

> - Stop the Docker ACAP application.

Reviewer 2: ```suggestion
- Uninstall the Docker ACAP application.
```

curl -s https://api.github.com/repos/AxisCommunications/docker-acap/releases/latest | grep "browser_download_url.*Docker_Daemon_.*_<ARCH>\_signed.eap"
```

#### Migrating from rootful Docker ACAP application

When migrating from a rootful Docker ACAP application, any version before 2.0, the following is recommended:

- Copy any images that you want to persist from the device to your computer.
- Stop the Docker ACAP application.
- Uninstall the Docker ACAP application.
- If you use the SD card as storage format it or manually remove the `dockerd` directory (`/var/spool/storage/SD-Disk/dockerd`).
- Restart the device.
- Install the rootless Docker ACAP application.

### Using TLS to secure the Docker ACAP application

> The Docker ACAP application can be run either unsecured or in TLS mode. The application uses

Reviewer 3: The Docker ACAP application can be run in either unsecured or TLS mode.

To get more informed about specifications, check the
[SD Card Standards][sd-card-standards].

> [!NOTE]
>
>If Docker ACAP application v1.4 or previous has been used on the device with SD card as storage
>the storage directory might already be created with root permissions.
>Since v2.0 the Docker ACAP application is run in rootless mode and it will then not be able
>to access that directory. To solve this, either reformat the SD card or manually

> >remove the directory that is used by the Docker ACAP application.

Reviewer 3: ```suggestion
>SSH into the device and remove the directory `dockerd` at `/var/spool/storage/SD_DISK/dockerd` that is used by the Docker ACAP application.
```

 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
    (arm64v8)
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run an Ubuntu container with:
 $ docker run -it ubuntu bash

Share images, automate workflows, and more with a free Docker ID:
 https://hub.docker.com/

For more examples and ideas, visit:
 https://docs.docker.com/get-started/

```

#### Loading images onto a device

If you have images in a local repository that you want to transfer to a device, or
if you have problems getting the `pull` command to work in your environment, `save`
and `load` can be used.

```sh
docker save <image on host local repository> | docker --tlsverify --host tcp://<device-ip>:$DOCKER_PORT load
```

#### Using host user secondary groups in container

The application is run by a non-root user on the device. This user is set
up to be a member in a number of secondary groups as listed in the

> [manifest.json](https://github.com/AxisCommunications/docker-acap/blob/main/app/manifest.json#L6-L11)

Author: Not sure about this link - maybe its better to refer to /app/manifest only

will be the default user inside the container. It will be mapped to the non-root user on
the device, and the group will be mapped to the non-root users primary group.
In order to get access inside the container to resources on the device that are group owned by any
of the non-root users secondary groups, these need to be added for the container user.
This can be done by using `group_add` in a docker-compose.yaml or `--group-add` if using the Docker cli.
Unfortunately, adding the name of a secondary group is not supported. Instead the *mapped* id
of the group need to be used. At the moment of writing this the mappings are:

| device group | container group id |
| ------------ | :----------------: |
| datacache    | "1"                |
| sdk          | "2"                |
| storage      | "3"                |
| vdo          | "4"                |
| optics       | "5"                |

Note that the names of the groups will *not* be correctly displayed inside the container.

## Building the Docker ACAP application

This repository provides a build script that uses Docker to build the application and then extract
the .eap file:

> ``` sh

Reviewer 2: ```suggestion
```sh
```
